### PR TITLE
feat(swarm): add `#[non_exhaustive]` to key enums

### PR DESCRIPTION
**Pull request by dhuseby**
## Description

Add `#[non_exhaustive]` to the following enums so that future additions don't cause breaking changes in downstream client code: `FromSwarm`, `ToSwarm`, `SwarmEvent`, `ConnectionHandlerEvent`, `ConnectionEvent`.

Related: #4543.

## Notes & open questions

This refactored a number of examples and protocols to have a default option when doing a `match` against these enums.

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: misc/allow-block-list/src/lib.rs
Comment (thomaseizinger) [inline misc/allow-block-list/src/lib.rs]: If we don't match on anything, we might as well not match at all. What do you think?

File: swarm/CHANGELOG.md
Comment (thomaseizinger) [inline swarm/CHANGELOG.md]:
```suggestion
- Add `#[non_exhaustive]` to `FromSwarm`, `ToSwarm`, `SwarmEvent`, `ConnectionHandlerEvent`, `ConnectionEvent`.
```

File: swarm/tests/swarm_derive.rs
Comment (thomaseizinger) [inline swarm/tests/swarm_derive.rs]: Same here, we can just ditch this now I'd say.

File: protocols/rendezvous/src/client.rs
Comment (thomaseizinger) [inline protocols/rendezvous/src/client.rs]: Just noticing this here. Continuing the loop here is a bug, we should return `Poll::Pending` from the function otherwise this function probably spins forever? Weird that we didn't notice this earlier.

Even better, we should probably use the `futures::ready!` macro.

Comment (thomaseizinger) [reply protocols/rendezvous/src/client.rs]: Would you be up for sending a separate PR for this?

**Review by thomaseizinger — COMMENTED** [review]
Thanks for getting this started!

File: protocols/rendezvous/src/client.rs
Comment (dhuseby) [reply protocols/rendezvous/src/client.rs]: Sure I can do that and fix this. I'm not sure what the `futures::read!` is. I'll have to look it up.

File: protocols/rendezvous/src/server.rs
Comment (dhuseby) [inline protocols/rendezvous/src/server.rs:230]: @thomaseizinger is this a bug as well? Should it return Poll::Pending in this case?

File: protocols/rendezvous/src/client.rs
Comment (thomaseizinger) [reply protocols/rendezvous/src/client.rs]: > Sure I can do that and fix this. I'm not sure what the `futures::read!` is. I'll have to look it up.

It is a macro for short-circuiting a call to a function that returns `Poll` and cascades up the `Poll::Pending` variant, a bit like `?` but for `Future`s.

File: protocols/rendezvous/src/server.rs
Comment (thomaseizinger) [reply protocols/rendezvous/src/server.rs:230]: It is yes, we should break out of the loop in case `self.inner.poll` returns `Poll::Pending`.

File: protocols/rendezvous/src/client.rs
Comment (dhuseby) [reply protocols/rendezvous/src/client.rs]: Do you have any example code that uses it?

File: protocols/rendezvous/src/server.rs
Comment (dhuseby) [reply protocols/rendezvous/src/server.rs:230]: this is not a bug because it just falls through to the `return Poll::Pending` a the bottom of the function.

File: protocols/rendezvous/src/client.rs
Comment (thomaseizinger) [reply protocols/rendezvous/src/client.rs]: Actually, nevermind. See line 297 in this diff? That is where we return `Poll::Pending`. GitHub hid that part of the diff so I didn't see it when making the comment :)

File: protocols/rendezvous/src/server.rs
Comment (thomaseizinger) [reply protocols/rendezvous/src/server.rs:230]: Right, thanks for digging into it!

File: protocols/rendezvous/src/client.rs
Comment (thomaseizinger) [inline protocols/rendezvous/src/client.rs:284]:
```suggestion
                Poll::Ready(_) => continue,
```

We should poll for more events instead.

File: swarm-derive/src/lib.rs
Comment (thomaseizinger) [inline swarm-derive/src/lib.rs]: So this is tricky right. This should never happen because we should _always_ update this crate to generate all possible match arms. As a result, I think we should use `unreachable!` here. If we hit this, this is a bug.

File: swarm-derive/src/lib.rs
Comment (thomaseizinger) [inline swarm-derive/src/lib.rs:1]: I think similarly to the other comment, we should panic on line 887 instead of ignoring all other events.

File: misc/allow-block-list/src/lib.rs
Comment (mxinden) [inline misc/allow-block-list/src/lib.rs]: Can we somehow make use of [`non_exhaustive_omitted_patterns_lint`](https://github.com/rust-lang/rust/issues/89554) here even though it is nightly only at this point? When we add a new `FromSwarm` variant, I find it useful to explicitly have to handle it in each of the `libp2p-xxx` crates in this monorepo. Given that each `libp2p-xxx` has to depend on the latest `libp2p-swarm` (due to workspace dependencies), this should not lead to any version mismatches.

Comment (thomaseizinger) [reply misc/allow-block-list/src/lib.rs]: Yeah we definitely can. Just have to keep using nightly Rust in CI. We just need to add it to the `[workspace.lints]` table in the root `Cargo.toml`.

Comment (thomaseizinger) [reply misc/allow-block-list/src/lib.rs]: I am in-different to be honest. I don't see a big gain in adopting this lint. It shouldn't block this PR as we can always make this change later as well.

File: protocols/rendezvous/src/client.rs
Comment (thomaseizinger) [inline protocols/rendezvous/src/client.rs]: Instead of adding this here, we need to extend the branch above that calls `map_out`.

File: swarm-derive/src/lib.rs
Comment (thomaseizinger) [reply swarm-derive/src/lib.rs]: I think we can actually cleanly solve this using `ToSwarm::map_in`. That function is defined within `libp2p-swarm` and will those correctly map all variants. And instead of `generate_event_match_arm`, we can probably use `map_out`.

So it should be something like:

```rust
#trait_to_impl::poll(&mut self.#field, cx).map_out(|event| #into_out_event).map_in(|event| #wrapped_event)
```

Perhaps even cleaner would be to generate the closures directly because then the variable names are co-located.

File: swarm-derive/Cargo.toml
Comment (thomaseizinger) [inline swarm-derive/Cargo.toml]: `0.34.0` is already an unreleased version, you don't need to bump it :)

File: swarm/Cargo.toml
Comment (thomaseizinger) [inline swarm/Cargo.toml]: No need for specifying versions for `dev-dependencies`.

Comment (thomaseizinger) [reply swarm/Cargo.toml]:
```suggestion
libp2p-swarm-derive = { path = "../swarm-derive" }                  # Using `path` here because this is a cyclic dev-dependency which otherwise breaks releasing.
```

**Review by thomaseizinger — COMMENTED** [review]
A few more things, otherwise LGTM!

File: misc/allow-block-list/src/lib.rs
Comment (thomaseizinger) [reply misc/allow-block-list/src/lib.rs]: @mxinden Adopting this lint will require us to permanently lint with nightly in CI whereas at the moment, we will only lint with nightly until 1.75 is out.

I think for the moment, just doing a catch-all is simpler. We can change this at any point so I'd suggest to not block this PR on activating this lint but rather ship it as a follow-up :)

> I find it useful to explicitly have to handle it in each of the `libp2p-xxx` crates in this monorepo.

I'd like to challenge the usefulness of this. It might just be me but having to extend 30+ `match` statements with a new variant is a task where my brain shuts off instead of carefully thinking about each one whether or not I want to handle this variant here. When I introduce a new variant, there a few select places I want to handle it in and having to fix the entire workspace before I can have a clean `cargo clippy --all-targets --all-features` just blocks the flow of feature development.

To move forward with this (and unblock the `v0.53` release) we need to either add a catch-all variant to all these `match` blocks or remove them altogether. Without activating the lint, they are functionally equivalent and the latter is already implement here, hence I think deferring this is a) safe and b) faster.

File: swarm-derive/Cargo.toml
Comment (dhuseby) [reply swarm-derive/Cargo.toml]: I fixed the version bump. I'm not entirely sure what is the correct solution for `protocols/rendezvous/src/client.rs` The #into_out_event isn't in scope where the #generate_event_match_arm is. I'm not familiar with this code so I'm not very confident in making this change without more help from you.

File: swarm-derive/src/lib.rs
Comment (mxinden) [reply swarm-derive/src/lib.rs]: That is a neat solution leveraging that we can exhaustively match in `libp2p-swarm`.

File: misc/allow-block-list/src/lib.rs
Comment (mxinden) [reply misc/allow-block-list/src/lib.rs]: :+1: sounds good to me. Thanks for the additional reasoning.

File: swarm-derive/Cargo.toml
Comment (thomaseizinger) [reply swarm-derive/Cargo.toml]: I'll fix it up tomorrow :)

File: Cargo.toml
Comment (mxinden) [inline Cargo.toml]: The `=` is a hard requirement from now on, correct? Can you please document this here?

```suggestion
# `libp2p-swarm-derive` may not be compatible with different `libp2p-swarm` non-breaking releases. E.g. `libp2p-swarm` might introduce a new enum variant `FromSwarm` (which is `#[non-exhaustive]`) in a non-breaking release. Older versions of `libp2p-swarm-derive` would not forward this enum variant within the `NetworkBehaviour` hierarchy. Thus the version pinning is required. 
libp2p-swarm-derive = { version = "=0.34.0", path = "swarm-derive" }
```

**Review by mxinden — COMMENTED** [review]
Looking good to me overall. Suggestion from @thomaseizinger on `map_in` sounds worth blocking on.

File: misc/allow-block-list/src/lib.rs
Comment (dhuseby) [reply misc/allow-block-list/src/lib.rs]: So does this lint just warn whenever there is a match on a non-exhaustive enum and not all branches are explicitly handled? The #[non_exhaustive] attribute just requires that any match has a catch-all match arm and then this lint forces a warning when that is the case to alert the developer that they should review it? I'm worried about consistent false positives causing developers to ignore warnings.

File: swarm-derive/Cargo.toml
Comment (dhuseby) [reply swarm-derive/Cargo.toml]: Please propose the change here so that I can learn about this code. I'll review and land it here.

**Comment by dhuseby** [comment]
@mxinden and @thomaseizinger should we consider adding #[non_exhaustive] to any of the error enums in swarm/src/lib.rs? How often do we add new errors? I'm looking at:
* DialError: https://github.com/libp2p/rust-libp2p/blob/d0f62e9e23b883b31c4720e509ad37e9c1709bab/swarm/src/lib.rs#L1489
* ListenError: https://github.com/libp2p/rust-libp2p/blob/d0f62e9e23b883b31c4720e509ad37e9c1709bab/swarm/src/lib.rs#L1591

File: misc/allow-block-list/src/lib.rs
Comment (dhuseby) [reply misc/allow-block-list/src/lib.rs]: I talked with @mxinden to get the answers to my questions. I was correct in my understanding. I think we should not add this simply because of the unbounded time we'll be dependent on the nightly toolchain for CI/CD.

**Comment by thomaseizinger** [comment]
For now, I'd suggest to stick to the enums we have and gain some experience with `non_exhaustive`.

File: swarm-derive/Cargo.toml
Comment (thomaseizinger) [reply swarm-derive/Cargo.toml]: Only saw this comment now sorry. I pushed up two commits already, hope you don't mind :)

Happy to explain in more details on what happens there!

File: swarm-derive/src/lib.rs
Comment (thomaseizinger) [inline swarm-derive/src/lib.rs:711]: @dhuseby The main idea is that we can delegate into `libp2p-swarm` here to map the individual variants for us which will enforce an exhaustive matching.

**Comment by thomaseizinger** [comment]
@mxinden This is ready for a review. Feel free to merge if you think it is ready.

**Review by thomaseizinger — APPROVED** [review]
Approving to signal that it is ready from my end but I also wrote some of the code :)

File: Cargo.toml
Comment (thomaseizinger) [inline Cargo.toml]: Ideally, we move this comment onto the same line. If the comment is above, sorting the lines alphabetically doesn't work as easily.

File: swarm-derive/src/lib.rs
Comment (dhuseby) [reply swarm-derive/src/lib.rs:711]: that's a lot simpler. this is slick.

File: Cargo.toml
Comment (dhuseby) [reply Cargo.toml]: I moved the comment to the same line.

**Comment by dhuseby** [comment]
All changes incorporated and merged up to master:HEAD.

**Comment by dhuseby** [comment]
I'm not sure why so many tests are failing. Any ideas?

**Comment by dhuseby** [comment]
ensure manifest and CHANGELOG are properly updated : /

**Comment by dhuseby** [comment]
I'll go through and update CHANGELOG's

**Comment by dhuseby** [comment]
@thomaseizinger pinning the version of swarm-derive is causing the package version check to fail because "0.34.0" != "=0.34.0":

```
Package version: 0.34.0
Specified version: =0.34.0
```

I suspect that this test should be modified to enforce that the swarm-derive version is pinned (e.g. "=semver") and then check that the pinned version matches.

**Comment by thomaseizinger** [comment]
> ensure manifest and CHANGELOG are properly updated : /

Ugh, sorry you ran into this! We've recently added this check and it is sometimes a bit noisy :sweat_smile: 
We have an escape-hatch to make these checks pass without having to clutter the changelogs with implementation details. It is the `internal-change` label. I'll apply it. Mind reverting the changelog commit again? It doesn't really matter to our users.

File: .github/workflows/ci.yml
Comment (thomaseizinger) [inline .github/workflows/ci.yml]: I think instead adding another check, we could also just use `||`:

```shell
test "$PACKAGE_VERSION" = "$SPECIFIED_VERSION" || test "=$PACKAGE_VERSION" = "$SPECIFIED_VERSION"
```

What do you think?

Comment (thomaseizinger) [reply .github/workflows/ci.yml]: That is a bit more general than we specifically need but semantically the same. All this check should enforce is that we keep the versions in sync. If we ever end up pinning another crate, it should likely apply there too!

**Comment by thomaseizinger** [comment]
CI should hopefully pass now. @mxinden please review :)

File: .github/workflows/ci.yml
Comment (dhuseby) [reply .github/workflows/ci.yml]: sure, I'll fix this.

Comment (dhuseby) [reply .github/workflows/ci.yml]: nvrmnd, you did it already.

**Review by mxinden — APPROVED** [review]
Great to see this landing.

Thanks @dhuseby for the initial suggestion.

Thanks @dhuseby and @thomaseizinger for the implementation.

**Comment by dhuseby** [comment]
> Great to see this landing.

Thank you @mxinden for you reviews!

**Comment by nazar-pc** [comment]
This is actually an annoying change for our request-response wrapper where we have kind of a factory protocol that creates other protocols internally and now that enum is non-exhaustive we have no compile-time guarantees about what variants are there, so during upgrade we can forget to forward some.

**Comment by thomaseizinger** [comment]
If you are okay with running lints on nightly, you could make use of https://github.com/rust-lang/rust/issues/89554.

**Comment by thomaseizinger** [comment]
For `ToSwarm`, I'd recommend to use the various `map` functions like `map_in` and `map_out`. Delegating to those will ensure you are exhaustively matching because they are defined in `libp2p-swarm`.

For `FromSwarm`, you should be able to forward the entire message without matching on it.

**Comment by nazar-pc** [comment]
Didn't know about map functions, might be helpful here, thanks!

**Comment by thomaseizinger** [comment]
> Didn't know about map functions, might be helpful here, thanks!

That is how we solve this issue for the custom-derive which also has to generate code that lives outside of `libp2p-swarm` but should always match exhaustively.

**Comment by thomaseizinger** [comment]
I just opened https://github.com/libp2p/rust-libp2p/pull/4825 which should help as well.

**Pull request merged**
